std/unicode.zig

```diff
@@ -17,12 +17,30 @@ pub fn utf8CodepointSequenceLength(c: u32) !u3 {
 /// Given the first byte of a UTF-8 codepoint,
 /// returns a number 1-4 indicating the total length of the codepoint in bytes.
 /// If this byte does not match the form of a UTF-8 start byte, returns Utf8InvalidStartByte.
-pub fn utf8ByteSequenceLength(first_byte: u8) !u3 {
-    if (first_byte < 0b10000000) return u3(1);
-    if (first_byte & 0b11100000 == 0b11000000) return u3(2);
-    if (first_byte & 0b11110000 == 0b11100000) return u3(3);
-    if (first_byte & 0b11111000 == 0b11110000) return u3(4);
-    return error.Utf8InvalidStartByte;
+const Utf8ByteSequenceLengthError = error{
+    Utf8InvalidStartByte,
+};
+pub fn utf8ByteSequenceLength(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
+    var len: u8 = undefined;
+    var clz: u4 = undefined;
+    const INVALID = u8(9);
+    clz = @clz(~first_byte);
+    len = switch (clz) {
+        0 => u8(1), // ASCII
+        1 => INVALID, // Continuation byte
+        2 => u8(2),
+        3 => u8(3),
+        4 => u8(4),
+        5 => u8(5),
```

Comment:
But that way llvm can't optimize the switch into a look-up table. I was specifically targeting that llvm optimization:

https://zig.godbolt.org/z/c1AQfs

Reply:
does this work? (godbolt doesn't work on mobile, and i can't really read x86 assembly anyway.)

```zig
    const leading_ones = @clz(~first_byte);
    if (leading_ones == 1 or leading_ones > 4) return -1;
    return leading_ones;
```

Reply:
That be two branches anyways (== 1 and > 4). My code uses only one branch.

Reply:
Write the code to express intent. If you want to write code that deviates from that, you have to measure performance, not just look at the assembly. Here's a benchmark:

```zig
const std = @import("std");

const Utf8ByteSequenceLengthError = error{Invalid};

pub fn utf8ByteSequenceLengthOptimized(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    var clz = @clz(~first_byte);
    var len = switch (clz) {
        0 => u8(1), // ASCII
        1 => u8(9), // Continuation byte
        2 => u8(2),
        3 => u8(3),
        4 => u8(4),
        5 => u8(5),
        6 => u8(6),
        7 => u8(7),
        8 => u8(8),
        else => unreachable,
    };
    if (len > 4) {
        return error.Invalid;
    }
    return @intCast(u3, len);
}

pub fn utf8ByteSequenceLengthSimple(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    switch (@clz(~first_byte)) {
        0 => return 1, // ASCII
        1 => return error.Invalid,
        2 => return 2,
        3 => return 3,
        4 => return 4,
        else => return error.Invalid,
    }
}

pub fn main() !void {
    const Item = struct {
        func: fn (num: u8) Utf8ByteSequenceLengthError!u3,
        name: []const u8,
    };
    const items = []Item{
        Item{
            .func = utf8ByteSequenceLengthOptimized,
            .name = "optimized",
        },
        Item{
            .func = utf8ByteSequenceLengthSimple,
            .name = "simple",
        },
        Item{
            .func = utf8ByteSequenceLengthOptimized,
            .name = "optimized",
        },
        Item{
            .func = utf8ByteSequenceLengthSimple,
            .name = "simple",
        },
    };
    const seed = 0x1234;
    // first make sure the values are all equal
    {
        var prng = std.rand.DefaultPrng.init(seed);
        var i: usize = 0;
        var sum: usize = 0;
        while (i < 10000000) : (i += 1) {
            const rand_byte = prng.random.int(u8);
            const opt_result = if (utf8ByteSequenceLengthOptimized(rand_byte)) |x| usize(x) else |_| 100;
            const simple_result = if (utf8ByteSequenceLengthSimple(rand_byte)) |x| usize(x) else |_| 100;
            if (opt_result != simple_result) {
                std.debug.panic("bad value: {}\n", rand_byte);
            }
        }
    }
    for (items) |item| {
        var prng = std.rand.DefaultPrng.init(seed);
        var timer = try std.os.time.Timer.start();
        var i: usize = 0;
        var sum: usize = 0;
        const start = timer.lap();
        while (i < 10000000) : (i += 1) {
            const rand_byte = prng.random.int(u8);
            if (item.func(rand_byte)) |x| {
                sum += x;
            } else |_| {
                sum += 100;
            }
        }
        const end = timer.read();
        const elapsed_s = @intToFloat(f64, end - start) / std.os.time.ns_per_s;
        const bytes_per_sec = @floatToInt(usize, @intToFloat(f64, i) / elapsed_s);
        std.debug.warn("{}: {Bi1}/sec, checksum: {}\n", item.name, bytes_per_sec, sum);
    }
}
```

```
$ ./zig build-exe test.zig 
$ ./test
optimized: 8.0MiB/sec, checksum: 291975483
simple: 8.9MiB/sec, checksum: 291975483
optimized: 8.0MiB/sec, checksum: 291975483
simple: 8.9MiB/sec, checksum: 291975483
$ ./zig build-exe test.zig --release-fast
$ ./test
optimized: 69.2MiB/sec, checksum: 291975483
simple: 69.3MiB/sec, checksum: 291975483
optimized: 69.4MiB/sec, checksum: 291975483
simple: 69.5MiB/sec, checksum: 291975483
```

So in debug mode the simple version is marginally faster, and in release-fast mode, they're indistinguishable.

Write code to express intent.

Even if this benchmark had produced results showing clearly that the non-simple one was faster, it would be better to put effort towards improving LLVM optimization passes to recognize a pattern, rather than to modify zig code to work around the implementation details of the optimizer.

Reply:
The benchmark is significant if you take the PRNG out of the loop:
```
optimized: 175.2MiB/sec, checksum: 291975483
simple: 124.9MiB/sec, checksum: 291975483
optimized: 171.3MiB/sec, checksum: 291975483
simple: 126.9MiB/sec, checksum: 291975483

const std = @import("std");
const Utf8ByteSequenceLengthError = error{
    Utf8InvalidStartByte,
};
pub fn utf8ByteSequenceLengthOptimized(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    const INVALID = 0;
    const swtch = []u8{1, INVALID, 2, 3, 4, INVALID, INVALID, INVALID, INVALID};
    var len = swtch[@clz(~first_byte)];
    if (len == INVALID) {
        return error.Utf8InvalidStartByte;
    }
    return @intCast(u3, len);
}

pub fn utf8ByteSequenceLengthSimple(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    if (first_byte < 0b10000000) return u3(1);
    if (first_byte & 0b11100000 == 0b11000000) return u3(2);
    if (first_byte & 0b11110000 == 0b11100000) return u3(3);
    if (first_byte & 0b11111000 == 0b11110000) return u3(4);
    return error.Utf8InvalidStartByte;
}

const Allocator = std.mem.Allocator;

pub fn main() !void {
    const Item = struct {
        func: fn (num: u8) Utf8ByteSequenceLengthError!u3,
        name: []const u8,
    };
    const items = []Item{
        Item{
            .func = utf8ByteSequenceLengthOptimized,
            .name = "optimized",
        },
        Item{
            .func = utf8ByteSequenceLengthSimple,
            .name = "simple",
        },
        Item{
            .func = utf8ByteSequenceLengthOptimized,
            .name = "optimized",
        },
        Item{
            .func = utf8ByteSequenceLengthSimple,
            .name = "simple",
        },
    };
    const seed = 0x1234;
    const allocator = &std.heap.DirectAllocator.init().allocator;
    const d = try allocator.alloc(u8, 10000000);
    {
        var prng = std.rand.DefaultPrng.init(seed);
        var i: usize = 0;
        var sum: usize = 0;
        while (i < 10000000) : (i += 1) {
            d[i] = prng.random.int(u8);
        }
    }
    // first make sure the values are all equal
    {
        var i: usize = 0;
        var sum: usize = 0;
        while (i < 10000000) : (i += 1) {
            const opt_result = if (utf8ByteSequenceLengthOptimized(d[i])) |x| usize(x) else |_| 100;
            const simple_result = if (utf8ByteSequenceLengthSimple(d[i])) |x| usize(x) else |_| 100;
            if (opt_result != simple_result) {
                std.debug.panic("bad value: {}\n", d[i]);
            }
        }
    }
    for (items) |item| {
        var timer = try std.os.time.Timer.start();
        var i: usize = 0;
        var sum: usize = 0;
        const start = timer.lap();
        while (i < 10000000) : (i += 1) {
            if (item.func(d[i])) |x| {
                sum += x;
            } else |_| {
                sum += 100;
            }
        }
        const end = timer.read();
        const elapsed_s = @intToFloat(f64, end - start) / std.os.time.ns_per_s;
        const bytes_per_sec = @floatToInt(usize, @intToFloat(f64, i) / elapsed_s);
        std.debug.warn("{}: {Bi1}/sec, checksum: {}\n", item.name, bytes_per_sec, sum);
    }
}
```

Reply:
OK great. Definitely we want the `utf8ByteSequenceLengthOptimized` that you have here.

For what it's worth, I actually think that both of these functions are equally readable:

```zig
pub fn utf8ByteSequenceLengthOptimized(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    const INVALID = 0;
    const swtch = []u8{1, INVALID, 2, 3, 4, INVALID, INVALID, INVALID, INVALID};
    var len = swtch[@clz(~first_byte)];
    if (len == INVALID) {
        return error.Utf8InvalidStartByte;
    }
    return @intCast(u3, len);
}

pub fn utf8ByteSequenceLengthSimple(first_byte: u8) Utf8ByteSequenceLengthError!u3 {
    if (first_byte < 0b10000000) return u3(1);
    if (first_byte & 0b11100000 == 0b11000000) return u3(2);
    if (first_byte & 0b11110000 == 0b11100000) return u3(3);
    if (first_byte & 0b11111000 == 0b11110000) return u3(4);
    return error.Utf8InvalidStartByte;
}
```